odoo/addons/base/models/ir_attachment.py

```diff
@@ -431,7 +431,13 @@ def check(self, mode, values=None):
             return True
         # Always require an internal user (aka, employee) to access to a attachment
         if not (self.env.is_admin() or self.env.user.has_group('base.group_user')):
-            raise AccessError(_("Sorry, you are not allowed to access this document."))
+            has_access = False
+            if self.ids:
+                self.env['ir.attachment'].flush(['public', 'create_uid'])
+                self._cr.execute('SELECT public, create_uid FROM ir_attachment WHERE id IN %s', [tuple(self.ids)])
+                has_access = all(public or create_uid == self.env.uid for public, create_uid in self._cr.fetchall())
+            if not has_access:
+                raise AccessError(_("Sorry, you are not allowed to access this document."))
```

Comment:
Alls calls: read (through binary_content), create, search are accessing the attachment through sudo.

If this change is necessary just for generate_access_token, we should probably not do this and instead do something like `.sudo().generate_access_token` in `_attachment_create` if we are not creating a `public` attachment.